aws/resource_aws_iot_thing_type.go

```diff
@@ -0,0 +1,186 @@
+		},
+	}
+
+	log.Printf("[DEBUG] Creating IoT Thing Type: %s", params)
+	out, err := conn.CreateThingType(params)
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*out.ThingTypeName)
+
+	return resourceAwsIotThingTypeUpdate(d, meta)
+}
+
+func resourceAwsIotThingTypeRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iotconn
+
+	params := &iot.DescribeThingTypeInput{
+		ThingTypeName: aws.String(d.Id()),
+	}
+	log.Printf("[DEBUG] Reading IoT Thing Type: %s", params)
+	out, err := conn.DescribeThingType(params)
+
+	if err != nil {
+		if isAWSErr(err, iot.ErrCodeResourceNotFoundException, "") {
+			log.Printf("[WARN] IoT Thing Type %q not found, removing from state", d.Id())
+			d.SetId("")
+		}
+		return err
+	}
+
+	if out.ThingTypeMetadata != nil {
+		d.Set("deprecated", out.ThingTypeMetadata.Deprecated)
+	}
+
+	d.Set("arn", *out.ThingTypeArn)
+	properties := out.ThingTypeProperties
+	if properties != nil {
+		d.Set("searchable_attributes", flattenStringList(properties.SearchableAttributes))
+
+		if properties.ThingTypeDescription != nil {
+			d.Set("description", *properties.ThingTypeDescription)
+		}
+	}
+
+	return nil
+}
+
+func resourceAwsIotThingTypeUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iotconn
+
+	if d.HasChange("deprecated") {
+		params := &iot.DeprecateThingTypeInput{
+			ThingTypeName: aws.String(d.Id()),
+		}
+
+		if v := d.Get("deprecated").(bool); !v {
+			params.UndoDeprecate = aws.Bool(true)
+		}
+
+		log.Printf("[DEBUG] Updating IoT Thing Type: %s", params)
+		_, err := conn.DeprecateThingType(params)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceAwsIotThingTypeRead(d, meta)
+}
+
+func resourceAwsIotThingTypeDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iotconn
+
+	// In order to delete an IoT Thing Type, you must deprecate it first and wait
+	// at least 5 minutes.
+	deprecateParams := &iot.DeprecateThingTypeInput{
+		ThingTypeName: aws.String(d.Id()),
+	}
+	log.Printf("[DEBUG] Deprecating IoT Thing Type: %s", deprecateParams)
+	_, err := conn.DeprecateThingType(deprecateParams)
+
+	if err != nil {
+		return err
+	}
+
+	deleteParams := &iot.DeleteThingTypeInput{
+		ThingTypeName: aws.String(d.Id()),
+	}
+	log.Printf("[DEBUG] Deleting IoT Thing Type: %s", deleteParams)
+
+	return resource.Retry(6*time.Minute, func() *resource.RetryError {
+		_, err := conn.DeleteThingType(deleteParams)
+
+		if err != nil {
+			if isAWSErr(err, iot.ErrCodeInvalidRequestException, "Please wait for 5 minutes after deprecation and then retry") {
+				return resource.RetryableError(err)
+			}
+
+			// As the delay post-deprecation is about 5 minutes, it may have been
+			// deleted in between, thus getting a Not Found Exception.
+			if isAWSErr(err, iot.ErrCodeResourceNotFoundException, "") {
+				d.SetId("")
+				return nil
+			}
+
+			return resource.NonRetryableError(err)
+		}
+
+		d.SetId("")
```

Comment:
Nitpick: I know we have that in many resources still, but it's redundant. The ID is emptied automatically in the core.

See https://github.com/hashicorp/terraform/blob/096847c9f774bfb946de7413260b30f9f6041241/helper/schema/resource.go#L197-L206